pkg/util/voljournal.go

```diff
@@ -0,0 +1,383 @@
+*/
+
+type CSIJournal struct {
+	// csiDirectory is the name of the CSI volumes object map that contains CSI volume-name (or
+	// snapshot name) based keys
+	csiDirectory string
+
+	// CSI volume-name keyname prefix, for key in csiDirectory, suffix is the CSI passed volume name
+	csiNameKeyPrefix string
+
+	// Per Ceph volume (RBD/FS-subvolume) object map name prefix, suffix is the generated volume uuid
+	cephUUIDDirectoryPrefix string
+
+	// CSI volume-name key in per Ceph volume object map, containing CSI volume-name for which the
+	// Ceph volume was created
+	csiNameKey string
+
+	// source volume name key in per Ceph snapshot object map, containing Ceph source volume uuid
+	// for which the snapshot was created
+	cephSnapSourceKey string
+
+	// volume name prefix for naming on Ceph rbd or FS, suffix is a uuid generated per volume
+	namingPrefix string
+}
+
+var (
+	CSIVolumeKeyNames = CSIJournal{
+		csiDirectory:            "csi.volumes",
+		csiNameKeyPrefix:        "csi.volume.",
+		cephUUIDDirectoryPrefix: "csi.volume.",
+		csiNameKey:              "csi.volname",
+		namingPrefix:            "csi-vol-",
+		cephSnapSourceKey:       "",
+	}
+
+	CSISnapshotKeyNames = CSIJournal{
+		csiDirectory:            "csi.snaps",
+		csiNameKeyPrefix:        "csi.snap.",
+		cephUUIDDirectoryPrefix: "csi.snap.",
+		csiNameKey:              "csi.snapname",
+		namingPrefix:            "csi-snap-",
+		cephSnapSourceKey:       "csi.source",
+	}
+)
+
+// CSIVolumeJournal returns an instance of volume keys
+func CSIVolumeJournal() CSIJournal {
+	return CSIVolumeKeyNames
+}
+
+// CSISnapshotSnapshot returns an instance of snapshot keys
+func CSISnapshotJournal() CSIJournal {
+	return CSISnapshotKeyNames
+}
+
+// NamingPrefix returns the value of naming prefix from the journal keys
+func (cj *CSIJournal) NamingPrefix() string {
+	return cj.namingPrefix
+}
+
+// SetCSIDirectorySuffix sets the given suffix for the csiDirectory omap
+func (cj *CSIJournal) SetCSIDirectorySuffix(suffix string) {
+	cj.csiDirectory = cj.csiDirectory + "." + suffix
+}
+
+/*
+CheckReservation checks if given request name contains a valid reservation
+- If there is a valid reservation, then the corresponding UUID for the volume/snapshot is returned
+- If there is a reservation that is stale (or not fully cleaned up), it is garbage collected using
+the UndoReservation call, as appropriate
+- If a snapshot is being checked, then its source is matched to the parentName that is provided
+
+NOTE: As the function manipulates omaps, it should be called with a lock against the request name
+held, to prevent parallel operations from modifying the state of the omaps for this request name.
+*/
+func (cj *CSIJournal) CheckReservation(monitors, id, key, pool, reqName, parentName string) (objUUID string, err error) {
+	var snapSource bool
+
+	if parentName != "" {
+		if cj.cephSnapSourceKey == "" {
+			err = errors.New("invalid snapshot request on volume journal")
+			return "", err
```

Comment:
I always get confused when I review  named return type instead of  `return " " , err` why can't we just do `return`

if we want to do `return "", err`   we have to declare named returns?

Reply:
Done (removed any further instances of naming return vars as well)